batch/test/test_batch.py

```diff
@@ -1777,3 +1793,417 @@ def test_region(client: BatchClient):
+    b.submit()
+
+    b_copy = client.get_batch(b.id)
+    jg_copy = b_copy.get_job_group(jg.job_group_id)
+    jg_copy.create_job(DOCKER_ROOT_IMAGE, ['true'])
+    b_copy.submit()
+    status = jg_copy.wait()
+    assert status['n_jobs'] == 1, str(status)
+
+
+def test_cancellation_doesnt_cancel_other_job_groups(client: BatchClient):
+    b = create_batch(client)
+    jg1 = b.create_job_group()
+    j1 = jg1.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    jg2 = b.create_job_group()
+    j2 = jg2.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    b.submit()
+
+    j1._wait_for_states('Running')
+
+    jg1.cancel()
+    jg1_status = jg1.wait()
+    jg2_status = jg2.status()
+
+    assert jg1_status['state'] == 'cancelled', str(jg1.debug_info())
+    assert jg2_status['state'] != 'cancelled', str(jg2.debug_info())
+
+    assert j1.status()['state'] == 'Cancelled', str(j1.status())
+    assert j2.status()['state'] != 'Cancelled', str(j2.status())
+
+    b.cancel()
+
+
+def test_dependencies_across_job_groups(client: BatchClient):
+    b = create_batch(client)
+    jg1 = b.create_job_group()
+    j1 = jg1.create_job(DOCKER_ROOT_IMAGE, ['true'])
+    jg2 = b.create_job_group()
+    jg2.create_job(DOCKER_ROOT_IMAGE, ['true'], parents=[j1])
+    b.submit()
+    status = b.wait()
+    assert status['state'] == 'success', str(b.debug_info())
+
+
+def test_job_group_cancel_after_n_failures_does_not_cancel_higher_up_jobs(client: BatchClient):
+    b = create_batch(client)
+    b_j = b.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    jg = b.create_job_group(cancel_after_n_failures=1)
+    jg.create_job(DOCKER_ROOT_IMAGE, ['false'])
+    j2 = jg.create_job(DOCKER_ROOT_IMAGE, ['sleep', '300'])
+    b.submit()
+    j2_status = j2.wait()
+    jg_status = jg.wait()
+    b_j_status = b_j.status()
+    try:
+        assert b_j_status['state'] != 'Cancelled', str((b_j_status, b.debug_info()))
+        assert j2_status['state'] == 'Cancelled', str((j2_status, jg.debug_info()))
+        assert jg_status['state'] == 'failure', str((jg_status, jg.debug_info()))
+    finally:
+        b.cancel()
+
+
+def test_cannot_create_job_in_job_group_that_has_been_cancelled(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    b.submit()
+    jg.cancel()
+    jg.create_job(DOCKER_ROOT_IMAGE, ['true'])
+    with pytest.raises(
+        httpx.ClientResponseError, match='bunch contains job where the job group has already been cancelled'
+    ):
+        b.submit()
+
+
+def test_cancellation_propogates_multiple_levels_top_down(client: BatchClient):
+    b = create_batch(client)
+    jg = b.create_job_group()
+    job_groups = [jg]
+    for _ in range(MAX_JOB_GROUP_NESTING_DEPTH - 1):
+        jg = jg.create_job_group()
+        job_groups.append(jg)
+    b.submit()
+    job_groups[0].cancel()
+
+    for jg in job_groups:
+        status = jg.status()
+        assert status['state'] == 'cancelled', str(status)
```

Comment:
Let's stick an attribute on the group that indicates it's level and let's include the debug_info of the batch in the assertion error message.

Does the job group status include attributes?

Reply:
It should.